transaction/transactor.go

```diff
@@ -15,57 +15,120 @@
  *
  */
 
-package common
+package transaction
 
 import (
 	"context"
 	"crypto/rand"
 	"encoding/base64"
 	"encoding/json"
-	"errors"
+	"io/ioutil"
 	"net"
 	"net/http"
 	"strings"
 	"time"
 
-	"github.com/xmidt-org/candlelight"
-
-	"github.com/xmidt-org/bascule"
-
 	kitlog "github.com/go-kit/kit/log"
-	"github.com/go-kit/kit/log/level"
 	kithttp "github.com/go-kit/kit/transport/http"
+	"github.com/xmidt-org/bascule"
+	"github.com/xmidt-org/candlelight"
 	"github.com/xmidt-org/webpa-common/v2/logging"
 )
 
-type transactionRequest struct {
+// HeaderWPATID is the header key for the WebPA transaction UUID
+const HeaderWPATID = "X-WebPA-Transaction-Id"
+
+// XmidtResponse represents the data that a tr1d1um transactor keeps from an HTTP request to
+// the XMiDT API
+type XmidtResponse struct {
+
+	//Code is the HTTP Status code received from the transaction
+	Code int
+
+	//ForwardedHeaders contains all the headers tr1d1um keeps from the transaction
+	ForwardedHeaders http.Header
+
+	//Body represents the full data off the XMiDT http.Response body
+	Body []byte
+}
+
+// Tr1d1umTransactor performs a typical HTTP request but
+// enforces some logic onto the HTTP transaction such as
+// context-based timeout and header filtering
+// this is a common utility for the stat and config tr1d1um services
+type Tr1d1umTransactor interface {
+	Transact(*http.Request) (*XmidtResponse, error)
+}
+
+// Tr1d1umTransactorOptions include parameters needed to configure the transactor
+type Tr1d1umTransactorOptions struct {
+	//RequestTimeout is the deadline duration for the HTTP transaction to be completed
+	RequestTimeout time.Duration
+
+	//Do is the core responsible to perform the actual HTTP request
+	Do func(*http.Request) (*http.Response, error)
+}
+
+type tr1d1umTransactor struct {
+	RequestTimeout time.Duration
+	Do             func(*http.Request) (*http.Response, error)
+}
+
+type TransactionRequest struct {
 	Address string `json:"address,omitempty"`
 	Path    string `json:"path,omitempty"`
 	Query   string `json:"query,omitempty"`
 	Method  string `json:"method,omitempty"`
 }
 
-func (re *transactionRequest) MarshalJSON() ([]byte, error) {
-	return json.Marshal(re)
-}
-
 type transactionResponse struct {
 	Code    int         `json:"code,omitempty"`
 	Headers interface{} `json:"headers,omitempty"`
 }
 
-type GetLoggerFunc func(context.Context) kitlog.Logger
+func (re *TransactionRequest) MarshalJSON() ([]byte, error) {
+	return json.Marshal(re)
+}
 
 func (rs *transactionResponse) MarshalJSON() ([]byte, error) {
 	return json.Marshal(rs)
 }
 
-// HeaderWPATID is the header key for the WebPA transaction UUID
-const HeaderWPATID = "X-WebPA-Transaction-Id"
+func NewTr1d1umTransactor(o *Tr1d1umTransactorOptions) Tr1d1umTransactor {
```

Comment:
The transactor is the main thing being provided by this package, so I think `New()` by itself is okay.

```suggestion
func New(o *Options) T {
```